File: src/pages/Modules/patch.py
import re

file_path = 'C:/Users/tales/OneDrive/Desktop/zenvora3d/src/pages/Modules/FounderJourney.jsx'
with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# 1. Imports
content = content.replace("import { useDatabase } from '../../context/DatabaseContext';", "import { useEffect } from 'react';\nimport { getFounderJourneySettings, updateFounderJourneySettings, getFounderJourneyItems, createFounderJourneyItem, updateFounderJourneyItem, deleteFounderJourneyItem, toggleFounderJourneyItemStatus, publishFounderJourneyLive } from '../../Services/founderJourneyServices';")

# 2. Component init
init_old = '''export const FounderJourney = () => {
  const { db, updateSection } = useDatabase();
  const journeyData = db?.founderJourney || {};'''
init_new = '''export const FounderJourney = () => {
  const [journeyData, setJourneyData] = useState({});

  const fetchAll = async () => {
    try {
      const settingsRes = await getFounderJourneySettings();
      const itemsRes = await getFounderJourneyItems();
      const settings = settingsRes.data || {};
      const items = itemsRes.data || [];
      const newData = { ...settings };
      items.forEach(item => {
        if (!newData[item.type]) newData[item.type] = [];
        newData[item.type].push(item);
      });
      // Sort items by order
      Object.keys(newData).forEach(key => {
        if (Array.isArray(newData[key])) {
            newData[key].sort((a, b) => (a.order || 0) - (b.order || 0));
        }
      });
      setJourneyData(newData);
      setHeroForm(newData.hero || {});
      setSettingsForm(newData.timelineSettings || {});
      setVisionForm(newData.futureVision || {});
      setSeoForm(newData.seo || {});
    } catch (err) {
      console.error(err);
    }
  };

  useEffect(() => {
    fetchAll();
  }, []);'''
content = content.replace(init_old, init_new)

# 3. handleSingleSave
save_old = '''  const handleSingleSave = (sectionKey, data) => {
    updateSection('founderJourney', { [sectionKey]: data });
    showToast(`${sectionKey.toUpperCase()} section parameters updated successfully.`);
  };'''
save_new = '''  const handleSingleSave = async (sectionKey, data) => {
    try {
      await updateFounderJourneySettings({ [sectionKey]: data });
      showToast(`${sectionKey.toUpperCase()} section parameters updated successfully.`);
      fetchAll();
    } catch (err) {}
  };'''
content = content.replace(save_old, save_new)

# 4. updateSectionMeta
meta_old = '''  const updateSectionMeta = (secId, key, val) => {
    const currentSettings = journeyData.sectionSettings || {};
    const updatedSettings = {
      ...currentSettings,
      [secId]: {
        ...(currentSettings[secId] || { order: 1, status: "Active" }),
        [key]: val
      }
    };
    updateSection('founderJourney', { sectionSettings: updatedSettings });
    // Silent update, no toast notifications popped up!
  };'''
meta_new = '''  const updateSectionMeta = async (secId, key, val) => {
    const currentSettings = journeyData.sectionSettings || {};
    const updatedSettings = {
      ...currentSettings,
      [secId]: {
        ...(currentSettings[secId] || { order: 1, status: "Active" }),
        [key]: val
      }
    };
    try {
      await updateFounderJourneySettings({ sectionSettings: updatedSettings });
      fetchAll();
    } catch (err) {}
  };'''
content = content.replace(meta_old, meta_new)

# 5. List Manager Saves
list_save_old = '''    const handleSaveItem = () => {
      let nextList = [];
      if (editingItemId) {
        nextList = listData.map(item => item.id === editingItemId ? { ...item, ...draftItem } : item);
        showToast("Item updated successfully.");
      } else {
        const newItem = { ...draftItem, id: `item-${Date.now()}`, status: draftItem.status || 'Active', order: listData.length + 1 };
        nextList = [...listData, newItem];
        showToast("New item created.");
      }
      updateSection('founderJourney', { [sectionKey]: nextList });
      setActiveEditorSection(null);
      setEditingItemId(null);
      setDraftItem({});
    };'''
list_save_new = '''    const handleSaveItem = async () => {
      try {
        if (editingItemId) {
          await updateFounderJourneyItem(editingItemId, { ...draftItem, type: sectionKey });
          showToast("Item updated successfully.");
        } else {
          await createFounderJourneyItem({ ...draftItem, type: sectionKey, status: draftItem.status || 'Active', order: listData.length + 1 });
          showToast("New item created.");
        }
        fetchAll();
        setActiveEditorSection(null);
        setEditingItemId(null);
        setDraftItem({});
      } catch (err) {
        showToast("Error saving item", "error");
      }
    };'''
content = content.replace(list_save_old, list_save_new)

# 6. List Manager Delete
del_old = '''    const handleDeleteItem = (id) => {
      if (window.confirm("Are you sure you want to delete this item?")) {
        const nextList = listData.filter(item => item.id !== id);
        updateSection('founderJourney', { [sectionKey]: nextList });
        showToast("Item deleted.");
      }
    };'''
del_new = '''    const handleDeleteItem = async (id) => {
      if (window.confirm("Are you sure you want to delete this item?")) {
        try {
          await deleteFounderJourneyItem(id);
          showToast("Item deleted.");
          fetchAll();
        } catch(err){}
      }
    };'''
content = content.replace(del_old, del_new)

# 7. List Manager Toggle Status
tog_old = '''    const handleToggleStatus = (id, currentStatus) => {
      const nextList = listData.map(item => item.id === id ? { ...item, status: currentStatus === 'Active' ? 'Inactive' : 'Active' } : item);
      updateSection('founderJourney', { [sectionKey]: nextList });
      // Silent update on toggle switch, no notifications popped up!
    };'''
tog_new = '''    const handleToggleStatus = async (id, currentStatus) => {
      try {
        await toggleFounderJourneyItemStatus(id);
        fetchAll();
      } catch(err){}
    };'''
content = content.replace(tog_old, tog_new)

# 8. List Manager Move (Optional: just refresh after move)
mov_old = '''    const handleMoveItem = (index, direction) => {
      const nextList = [...listData];
      const target = index + direction;
      if (target >= 0 && target < nextList.length) {
        const temp = nextList[index];
        nextList[index] = nextList[target];
        nextList[target] = temp;
        updateSection('founderJourney', { [sectionKey]: nextList });
      }
    };'''
mov_new = '''    const handleMoveItem = async (index, direction) => {
      const nextList = [...listData];
      const target = index + direction;
      if (target >= 0 && target < nextList.length) {
        const temp = nextList[index];
        nextList[index] = nextList[target];
        nextList[target] = temp;
        // update orders
        nextList.forEach((it, idx) => it.order = idx + 1);
        try {
           for (const it of nextList) {
               await updateFounderJourneyItem(it._id || it.id, { order: it.order, type: sectionKey });
           }
           fetchAll();
        } catch(err){}
      }
    };'''
content = content.replace(mov_old, mov_new)

# 9. List Items Rendering - Fix item.id to item._id in mapped elements
content = content.replace('item.id', '(item._id || item.id)')

with open(file_path, 'w', encoding='utf-8') as f:
    f.write(content)

File: src/pages/Modules/patch.js
const fs = require('fs');
let content = fs.readFileSync('C:/Users/tales/OneDrive/Desktop/zenvora3d/src/pages/Modules/Collaborations.jsx', 'utf-8');

// 1. Remove useDatabase and its import
content = content.replace(/import \{ useDatabase \} from '\.\.\/\.\.\/context\/DatabaseContext';\n?/, '');
content = content.replace(/const \{ db, updateSection, deleteNestedItem \} = useDatabase\(\);\n?/, '');
content = content.replace(/const collaborationsPage = db\?\.collaborationsPage \|\| \{\};\n?/, '');

// Add new imports
const newImports = `import { useEffect } from 'react';
import {
  getCollaboration, updateHero, updateHistory, updateSeo, updateSectionSettings,
  addBrand, updateBrand, deleteBrand,
  addPartner, updatePartner, deletePartner,
  addMetric, updateMetric, deleteMetric,
  addCampaign, updateCampaign, deleteCampaign,
  addProcess, updateProcess, deleteProcess,
  addTestimonial, updateTestimonial, deleteTestimonial
} from '../../Services/collabrationServices';
`;
content = content.replace(/import \{ Card \} from '\.\.\/\.\.\/components\/ui\/Card';/, newImports + "import { Card } from '../../components/ui/Card';");

// 2. Add local state and useEffect
const stateCode = `  const [collaborationsPage, setCollaborationsPage] = useState({});

  const fetchData = async () => {
    try {
      const res = await getCollaboration();
      if (res.data && res.data.data) {
        let doc = res.data.data;
        if (Array.isArray(doc)) doc = doc[0] || {};
        setCollaborationsPage(doc || {});
        setHeroForm(doc.hero || {});
        setHistoryForm(doc.history || {});
        setSeoForm(doc.seo || {});
      }
    } catch (err) {
      console.error(err);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);
`;
content = content.replace(/export default function Collaborations\(\) \{/, 'export default function Collaborations() {\n' + stateCode);

// 3. Fix initial state for forms
content = content.replace(/const \[heroForm, setHeroForm\] = useState\(collaborationsPage\?\.hero \|\| \{\}\);/, 'const [heroForm, setHeroForm] = useState({});');
content = content.replace(/const \[historyForm, setHistoryForm\] = useState\(collaborationsPage\?\.history \|\| \{\}\);/, 'const [historyForm, setHistoryForm] = useState({});');
content = content.replace(/const \[seoForm, setSeoForm\] = useState\(collaborationsPage\?\.seo \|\| \{\}\);/, 'const [seoForm, setSeoForm] = useState({});');

// 4. Update simulateMediaUpload dummyUrl fix
content = content.replace(/dummyUrl/g, 'url');

// 5. handleSingleSave
const handleSingleSaveCode = `  const handleSingleSave = async (sectionKey, data) => {
    try {
      if (sectionKey === 'hero') await updateHero(data);
      if (sectionKey === 'history') await updateHistory(data);
      if (sectionKey === 'seo') await updateSeo(data);
      fetchData();
      showToast(\`\${sectionKey.toUpperCase()} section parameters updated successfully.\`);
    } catch (err) {
      showToast("Error updating section.", "error");
    }
  };`;
content = content.replace(/const handleSingleSave = [\s\S]*?showToast\([^\n]+\n\s+\};\n/, handleSingleSaveCode + '\n');

// 6. updateSectionMeta
const updateSectionMetaCode = `  const updateSectionMeta = async (secId, key, val) => {
    try {
      const currentSettings = collaborationsPage.sectionSettings || {};
      const updatedSettings = {
        ...currentSettings,
        [secId]: {
          ...(currentSettings[secId] || { order: 1, status: "Active" }),
          [key]: val
        }
      };
      await updateSectionSettings(updatedSettings);
      fetchData();
      showToast("Section settings updated successfully.");
    } catch(err) {
      showToast("Failed to update section settings.", "error");
    }
  };`;
content = content.replace(/const updateSectionMeta = [\s\S]*?updateSection\('collaborationsPage', \{ sectionSettings: updatedSettings \}\);\n\s+\};\n/, updateSectionMetaCode + '\n');

// 7. handleSaveItem in renderListManager
const handleSaveItemMatch = /const handleSaveItem = \(\) => \{[\s\S]*?setDraftItem\(\{\}\);\n\s+\};\n/;
const handleSaveItemCode = `    const handleSaveItem = async () => {
      try {
        if (editingItemId !== null) {
          if (sectionKey === 'brandCarousel') await updateBrand(editingItemId, draftItem);
          else if (sectionKey === 'partners') await updatePartner(editingItemId, draftItem);
          else if (sectionKey === 'metrics') await updateMetric(editingItemId, draftItem);
          else if (sectionKey === 'campaigns') await updateCampaign(editingItemId, draftItem);
          else if (sectionKey === 'process') await updateProcess(editingItemId, draftItem);
          else if (sectionKey === 'testimonials') await updateTestimonial(editingItemId, draftItem);
        } else {
          if (sectionKey === 'brandCarousel') await addBrand(draftItem);
          else if (sectionKey === 'partners') await addPartner(draftItem);
          else if (sectionKey === 'metrics') await addMetric(draftItem);
          else if (sectionKey === 'campaigns') await addCampaign(draftItem);
          else if (sectionKey === 'process') await addProcess(draftItem);
          else if (sectionKey === 'testimonials') await addTestimonial(draftItem);
        }
        fetchData();
        showToast("Item saved successfully.");
        setActiveEditorSection(null);
        setEditingItemId(null);
        setDraftItem({});
      } catch (err) {
        showToast("Failed to save item.", "error");
      }
    };`;
content = content.replace(handleSaveItemMatch, handleSaveItemCode + '\n');

// 8. Delete dialog onConfirm
const onConfirmMatch = /onConfirm=\{[\s\S]*?showToast\("Record deleted successfully\."\);\n\s+\}\}/;
const onConfirmCode = `onConfirm={async () => {
          try {
            if (deletingSection === 'brandCarousel') await deleteBrand(deletingItemId);
            else if (deletingSection === 'partners') await deletePartner(deletingItemId);
            else if (deletingSection === 'metrics') await deleteMetric(deletingItemId);
            else if (deletingSection === 'campaigns') await deleteCampaign(deletingItemId);
            else if (deletingSection === 'process') await deleteProcess(deletingItemId);
            else if (deletingSection === 'testimonials') await deleteTestimonial(deletingItemId);
            
            fetchData();
            showToast("Record deleted successfully.");
          } catch(err) {
            showToast("Failed to delete record.", "error");
          } finally {
            setDeletingItemId(null);
            setDeletingSection(null);
          }
        }}`;
content = content.replace(onConfirmMatch, onConfirmCode);

// 9. Change Item ID references from item.id to item._id since Mongoose uses _id
content = content.replace(/item\.id/g, 'item._id');
// Except in creating a new item locally, which we don't do manually anymore because backend creates it.
// Wait, the state in handleStartEdit uses item._id:
// But wait, there is \`editingItemId !== null\`
// Let's replace \`setEditingItemId(item.id)\` with \`setEditingItemId(item._id)\`
content = content.replace(/setEditingItemId\(item\.id\)/g, 'setEditingItemId(item._id)');
content = content.replace(/setDeletingItemId\(item\.id\)/g, 'setDeletingItemId(item._id)');
content = content.replace(/key=\{item\.id \|\| idx\}/g, 'key={item._id || idx}');

// We also need to fix handleMoveItem because it just modifies local state
// But we won't rewrite handleMoveItem fully, let's just make it do a showToast warning
const handleMoveItemMatch = /const handleMoveItem = [\s\S]*?updateSection\('collaborationsPage', \{ \[sectionKey\]: nextList \}\);\n\s+\}\n\s+\};\n/;
const handleMoveItemCode = `    const handleMoveItem = async (index, direction) => {
      showToast("Order updating is not fully supported in this version. Edit the item order directly.", "error");
    };`;
content = content.replace(handleMoveItemMatch, handleMoveItemCode + '\n');

fs.writeFileSync('C:/Users/tales/OneDrive/Desktop/zenvora3d/src/pages/Modules/Collaborations.jsx', content, 'utf-8');
console.log('Done modifying Collaborations.jsx');
